### PR TITLE
Add a script that splits a branch's diff into code vs imports, comments and blanks

### PR DESCRIPTION
**Pull request by stefan-burke**
## What this adds

A small developer tool that measures how much of a branch's change is actual logic versus documentation, imports, and blank lines.

Given a base ref (defaults to `origin/main`), it reads `git diff --unified=0 …HEAD` and reports the changed lines broken down by:

- **area** — `src` / `test` / `other`
- **kind** — code, import, comment, blank

It then prints the code-only src-vs-test comparison and their ratio.

The classifier tracks state across lines, so it counts as imports/comments (not code):

- multi-line `import { … } from …` blocks
- multi-line `export { … } from …` / `export type { … } from …` barrel re-exports
- `/* … */` block comments and their `*` continuation lines

Added and removed files are both counted against their real area — a deletion names its path only on the `--- a/…` header, so both file headers are read.

## How it's built

The logic lives in `scripts/diff-code-lines-lib.ts` as pure, data-in/data-out functions (classify a line, tally a diff, format the report) behind a thin CLI shell (`scripts/diff-code-lines.ts`) that only runs `git diff` and prints — the same split as the existing `line-counts` tool. `test/scripts/diff-code-lines.test.ts` covers the classifier and diff tally at 100% line and branch coverage.

## Why

Raw diff line counts overstate how much real code changed when a codebase documents heavily (as this one does). This makes it easy to see, for any branch, how much is new logic versus comments and tests.

## Usage

```bash
deno run --allow-run scripts/diff-code-lines.ts [baseRef]
```

The classifier is a line-based heuristic, not a full parser, so the numbers are a close estimate rather than an exact count. Its one known blind spot: with `--unified=0` the unchanged `import {` opener stays off-diff, so adding or removing a single member inside an existing multi-line import reads as code.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Deno command-line tool that analyzes `git diff` and outputs a formatted report of changed lines.
  * Reports added/removed counts split by area (source, tests, other) and by type (code, imports, comments, blank lines), including multi-hunk and rename handling.
  * Includes a test/source ratio in the report when source code changes are present.
* **Tests**
  * Added a comprehensive test suite covering multi-line import/comment parsing, misclassification edge cases, correct per-area add/remove tallies, and report formatting.

### CODE REVIEW

File: scripts/diff-code-lines.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  let area: Area = "other";
+  let addState = freshState();
+  let removeState = freshState();
+
+  for (const line of diff.split("\n")) {
+    if (line.startsWith("+++ b/")) {
+      area = areaOf(line.slice("+++ b/".length));
+      continue;
+    }
+    if (line.startsWith("--- ")) continue;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Fixed in ec73f1a. `fileHeaderArea` now reads both file headers and derives the area from `--- a/…` when the `+++` side is `/dev/null`, so a deleted file's removed lines count under its real area. Regression test: "attributes a deleted file's removed lines to its real area".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    // A new hunk breaks any open import/comment run from the previous hunk.
+    if (line.startsWith("@@")) {
+      addState = freshState();
+      removeState = freshState();
+      continue;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Acknowledged as a known limitation rather than fixed. With `--unified=0` the unchanged `import {` opener stays off-diff, so a single added/removed member is a bare identifier line indistinguishable from code without the file/AST. I've documented this explicitly as the classifier's one known blind spot in the module docstring rather than adding fragile hunk-header parsing. The tool is a rough estimate, not an exact count.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  if (line.startsWith("/*")) {
+    state.inBlockComment = opensBlock(line, "/*", "*/");
+    return "comment";
+  }
+  if (line === "") return "blank";
+  if (line.startsWith("//") || line.startsWith("*")) return "comment";
+
+  if (/^import\b/.test(line) || /^export\b.*\bfrom\b/.test(line)) {
+    state.inImport = opensBlock(line, "{", "}");
+    return "import";
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Fixed in ec73f1a. `isImportLine` now enters import state for an `export {` / `export type {` opener that starts a multi-line block, so a barrel re-export block counts as import across all its lines (the closing `} from "…"` closes the run). Regression test: "recognises imports and re-exports, single- and multi-line".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  if (open !== null) return open;
+
+  if (line.startsWith("/*")) {
+    state.inBlockComment = opensBlock(line, "/*", "*/");
+    return "comment";
+  }
+  if (line === "") return "blank";
+  if (line.startsWith("//") || line.startsWith("*")) return "comment";
+
+  if (/^import\b/.test(line) || /^export\b.*\bfrom\b/.test(line)) {
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Fixed in ec73f1a. The single-line re-export check now requires a quoted module name after `from` (`/^export\b.*\bfrom\s+["']/`), so `export const pick = (view, from) => …` and other code that merely mentions the word stay classified as code. Regression test: "does not mistake code that merely mentions `from` for an import".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  );
+  console.log(`  src : ${srcCode}`);
+  console.log(`  test: ${testCode}`);
+  if (srcCode > 0) {
+    console.log(`  test/src ratio: ${(testCode / srcCode).toFixed(2)}`);
+  }
+};
+
+const base = Deno.args[0] ?? "origin/main";
+report(tallyDiff(await runGitDiff(base)));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Done in ec73f1a. Following the `line-counts` precedent, the pure logic moved to `scripts/diff-code-lines-lib.ts` behind a thin CLI shell, and `test/scripts/diff-code-lines.test.ts` now covers `classify` (imports, single- and multi-line re-exports, block comments, blanks, state across continuations) and `tallyDiff` (added/removed area counts, deleted-file attribution, hunk-boundary resets) at 100% line and branch coverage.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  }
+  if (line === "") return "blank";
+  if (line.startsWith("//") || line.startsWith("*")) return "comment";
+
+  if (/^import\b/.test(line) || /^export\b.*\bfrom\b/.test(line)) {
+    state.inImport = opensBlock(line, "{", "}");
+    return "import";
+  }
+  return "code";
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Fixed in ec73f1a — `isImportLine` enters import state for an `export {` / `export type {` opener that starts a multi-line block, so barrel re-exports count as import across the whole block. I used the `export {` / `export type {` shape guarded by `opensBlock` rather than `^export\b\s*(\*|\{)` so that a multi-line `export const config = { … }` object literal isn't mistaken for an import. Regression test: "recognises imports and re-exports, single- and multi-line".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    if (line.startsWith("+")) {
+      added[area][classify(line.slice(1), addState)] += 1;
+    } else if (line.startsWith("-")) {
+      removed[area][classify(line.slice(1), removeState)] += 1;
+    }
+  }
+  return { added, removed };
+};
+
+const netCode = (t: Tally): number => t.code;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Done in ec73f1a — renamed `netCode` to `codeCount`, since it's a plain `Tally.code` accessor and the add+remove summation happens at the call site in `formatReport`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      addState = freshState();
+      removeState = freshState();
+      continue;
+    }
+    if (line.startsWith("+")) {
+      added[area][classify(line.slice(1), addState)] += 1;
+    } else if (line.startsWith("-")) {
+      removed[area][classify(line.slice(1), removeState)] += 1;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts]: Fixed in ec73f1a. Header handling is now scoped to the file-header section (between `diff --git` and the first `@@`) via an `inHeader` flag, and `fileHeaderArea` reads both sides so a deletion's `--- a/…` sets the area. This fixes both the deleted-file mis-attribution and the dropping of a dash-prefixed content line. Regression tests: "attributes a deleted file's removed lines to its real area" and "classifies a dash-prefixed content line instead of dropping it".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines-lib.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+ *  - a single-line `export … from "…"` re-export — the quoted module name after
+ *    `from` is what tells a real re-export apart from code that merely mentions
+ *    the word, like `export const pick = (view, from) => …`;
+ *  - an `export {` / `export type {` that opens a multi-line block, which barrel
+ *    files close with `} from "…"` on a later line (that closing line is counted
+ *    by the open-run handler, so the opener only needs to start the run). */
+const isImportLine = (line: string): boolean =>
+  /^import\b/.test(line) ||
+  /^export\b.*\bfrom\s+["']/.test(line) ||
+  (/^export\s+(type\s+)?\{/.test(line) && opensBlock(line, "{", "}"));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines-lib.ts]: Fixed in 31d0c51. `isImportLine` now requires a space or quote right after `import` (`/^import["'\s]/`), so `import.meta` and dynamic `import(…)` classify as code. A re-export is recognised only when it starts with `export {` / `export *` / `export type` (`startsReExport`), so `export const note = 'from "x"'` stays code even though its string contains "from". Regression tests: "treats import.meta and dynamic import() as code, not import" and the `export const note` case in "does not mistake code that merely mentions `from`".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines-lib.ts [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+  const state: WalkState = {
+    addState: freshState(),
+    area: "other",
+    counts: { added: emptyAreaTally(), removed: emptyAreaTally() },
+    inHeader: false,
+    removeState: freshState(),
+  };
+  for (const line of diff.split("\n")) {
+    if (!applyBoundaryLine(line, state)) tallyContentLine(line, state);
+  }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines-lib.ts]: Done in 31d0c51. The diff walk is now a curried `reduce` from `#fp` (`reduce(foldLine, freshWalkState())(diff.split("\n")).counts`), keeping `applyBoundaryLine`/`tallyContentLine` as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    `${base}...HEAD`,
+  ]);
+  if (!result.success) {
+    throw new Error(`git diff failed: ${result.stderr}`);
+  }
+  return result.stdout;
+};
+
+const base = Deno.args[0] ?? "origin/main";
+console.log(formatReport(tallyDiff(await runGitDiff(base))));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines.ts:33]: Leaving this as-is, matching the existing `scripts/line-counts.ts` precedent: a thin CLI shell that runs at module load (`await printLineCounts(["src", "test"], console.log)`) with no `main()` guard and no test — all its logic lives in `line-counts-lib.ts`, which is what the test covers. `diff-code-lines.ts` follows the same shape: the pure, testable logic is fully covered in `diff-code-lines-lib.ts` (100% line and branch), and the 3-line shell only wires `git diff` to `formatReport`. A `main()`/`import.meta.main` guard would only add value if something imported the shell, but nothing does or should — the lib is the import surface. Keeping the two CLI shells consistent is the reason to not diverge here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/diff-code-lines-lib.ts [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+ *    (both executable code) stay classified as code;
+ *  - a single-line `export … from "…"` re-export (the quoted module name is the
+ *    real module specifier);
+ *  - an `export {` / `export type {` that opens a multi-line block, which barrel
+ *    files close with `} from "…"` on a later line (that closing line is counted
+ *    by the open-run handler, so the opener only needs to start the run). */
+const isImportLine = (line: string): boolean =>
+  /^import["'\s]/.test(line) ||
+  (startsReExport(line) &&
+    (/\bfrom\s+["']/.test(line) || opensBlock(line, "{", "}")));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines-lib.ts:109]: Fixed the dynamic-import half in c932e30: `isImportLine` now excludes anything with `.` or `(` after `import` (any spacing) via `!/^import\s*[.(]/`, so `import (…)` with a space classifies as code too. Regression test added ("import ("./spaced.ts")"). The local `export { … }` half is the same trade-off as the sibling thread — see my reply there.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/diff-code-lines-lib.ts:109]: Leaving this as a documented limitation rather than fixing it. A multi-line local `export { … }` (no `from`) opens with a line identical to a barrel re-export's opener — `export {`. The earlier review pass explicitly asked for barrel re-exports (`export { … } from "…"` split across lines) to be counted as import, and this single-pass classifier decides each line as it reads it, so at the opener there is no way to know whether a `from` will arrive later. Distinguishing the two would require buffering the block's lines and retroactively reclassifying — a provisional-state mechanism that contradicts the streaming design, for a rare construct in a tool that is explicitly a rough estimate, not a parser.

I've documented it as a known blind spot in the module docstring (alongside the single-member import-churn case) rather than add that machinery. Note the single-line local `export { a };` is already correctly counted as code (its opener closes on the same line, so `opensBlock` is false). If this edge ever matters in practice, the honest fix is a real parser, not a heuristic patch.

---
_Generated by [Claude Code](https://claude.ai/code)_
